File: app/src/main/java/com/anokmik/tripassistant/databinding/components/TextViewComponent.java
package com.anokmik.tripassistant.databinding.components;

import android.databinding.BindingAdapter;
import android.text.Spanned;
import android.widget.TextView;

public final class TextViewComponent {

    @BindingAdapter("android:text")
    public void setText(TextView view, CharSequence text) {
        CharSequence oldText = view.getText();
        if (text != oldText && (text != null || oldText.length() != 0)) {
            if (text instanceof Spanned) {
                if (text.equals(oldText)) {
                    return;
                }
            } else if (!haveContentsChanged(text, oldText)) {
                return;
            }

            view.setText(text);
        }
    }

    private boolean haveContentsChanged(CharSequence str1, CharSequence str2) {
        if (str1 == null != (str2 == null)) {
            return true;
        } else if (str1 == null) {
            return false;
        } else {
            int length = str1.length();
            if (length != str2.length()) {
                return true;
            } else {
                for (int i = 0; i < length; ++i) {
                    if (str1.charAt(i) != str2.charAt(i)) {
                        return true;
                    }
                }

                return false;
            }
        }
    }

}

File: app/src/main/kotlin/com/anokmik/tripassistant/databinding/components/TextViewComponent.kt
package com.anokmik.tripassistant.databinding.components

import android.databinding.BindingAdapter
import android.text.Spanned
import android.widget.EditText
import android.widget.TextView

class TextViewComponent {

    @BindingAdapter("android:text")
    fun setText(view: TextView, text: CharSequence?) {
        val oldText = view.text
        if (text !== oldText && (text != null || oldText.length != 0)) {
            if (text is Spanned) {
                if (text == oldText) {
                    return
                }
            } else if (!haveContentsChanged(text, oldText)) {
                return
            }

            if (view is EditText) view.append(text) else view.setText(text)
        }
    }

    private fun haveContentsChanged(str1: CharSequence?, str2: CharSequence?): Boolean {
        if (str1 == null != (str2 == null)) {
            return true
        } else if (str1 == null) {
            return false
        } else if (str2 == null) {
            return true
        } else {
            val length = str1.length
            if (length != str2.length) {
                return true
            } else {
                for (i in 0..length - 1) {
                    if (str1[i] != str2[i]) {
                        return true
                    }
                }
                return false
            }
        }
    }

}